pmd-java/src/test/java/net/sourceforge/pmd/lang/java/oom/DataStructureTest.java

```diff
@@ -82,4 +126,123 @@ public void testAddField() {
+
+        acu.jjtAccept(new JavaParserVisitorReducedAdapter() {
+            @Override
+            public Object visit(ASTMethodOrConstructorDeclaration node, Object data) {
+                res.add((int) toplevel.compute(opMetricKey, node, true, Version.STANDARD));
+                return super.visit(node, data);
+            }
+
+
+            @Override
+            public Object visit(ASTAnyTypeDeclaration node, Object data) {
+                res.add((int) toplevel.compute(classMetricKey, node, true, Version.STANDARD));
+                return super.visit(node, data);
+            }
+        }, null);
+
+        System.out.println();
+
+        final List<Integer> cmp = new ArrayList<>();
+
+        acu.jjtAccept(new JavaParserVisitorReducedAdapter() {
+            @Override
+            public Object visit(ASTMethodOrConstructorDeclaration node, Object data) {
+                cmp.add((int) toplevel.compute(opMetricKey, node, false, Version.STANDARD));
+                return super.visit(node, data);
+            }
+
+
+            @Override
+            public Object visit(ASTAnyTypeDeclaration node, Object data) {
+                cmp.add((int) toplevel.compute(classMetricKey, node, false, Version.STANDARD));
+                return super.visit(node, data);
+            }
+        }, null);
+
+        assertEquals(res, cmp);
+
+    }
+
+
+    @Test
+    public void forceMemoizationTest() {
+
+        final List<Integer> res = new ArrayList<>();
+        final PackageStats toplevel = Metrics.getTopLevelPackageStats();
+
+        ASTCompilationUnit acu = parseAndVisitForClass15(MetricsVisitorTestData.class);
+
+        acu.jjtAccept(new JavaParserVisitorReducedAdapter() {
+            @Override
+            public Object visit(ASTMethodOrConstructorDeclaration node, Object data) {
+                res.add((int) toplevel.compute(opMetricKey, node, true, Version.STANDARD));
+                return super.visit(node, data);
+            }
+
+
+            @Override
+            public Object visit(ASTAnyTypeDeclaration node, Object data) {
+                res.add((int) toplevel.compute(classMetricKey, node, true, Version.STANDARD));
+                return super.visit(node, data);
+            }
+        }, null);
+
+        System.out.println();
+
+        final List<Integer> cmp = new ArrayList<>();
+
+        acu.jjtAccept(new JavaParserVisitorReducedAdapter() {
+            @Override
+            public Object visit(ASTMethodOrConstructorDeclaration node, Object data) {
+                cmp.add((int) toplevel.compute(opMetricKey, node, true, Version.STANDARD));
+                return super.visit(node, data);
+            }
+
+
+            @Override
+            public Object visit(ASTAnyTypeDeclaration node, Object data) {
+                cmp.add((int) toplevel.compute(classMetricKey, node, true, Version.STANDARD));
+                return super.visit(node, data);
+            }
+        }, null);
```

Comment:
Looks like 4 of these anonymous classes?  Do they differ besides the `List` they populate and the `boolean` parameter?  The nature of the tests might be more obvious if there was one class (or factory method?) with whose constructor took those two arguments, and then used 4 times by these two tests.  I suspect the tests are really about verifying behavior for interesting combinations of variations of those two parameters.

Without doing this, there's 2 fairly complex test cases spread over nearly 100 lines.  Intent is difficult to discern, due to the size and degree of code analysis needed.